gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -156,33 +167,118 @@ public void handleResponse(WriteObjectResponse response) {
+    int bytesWritten = 0;
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    // No point in writing further if upload failed on another thread.
+    if (uploadOperation.isDone()) {
+      waitForCompletionAndThrowIfUploadFailed();
+    }
+
+    // create chunks of MAX_BYTES_PER_MESSAGE and push them into writeBuffer. If the loop ends up
+    // in a scenario where the lastChunk size is less than MAX_BYTES_PER_MESSAGE, then we hold on
+    // to it until next write or close.
+    while (buffer.remaining() > 0) {
+      int bytesToWrite = 0;
+      if (lastChunk != null) {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE - lastChunk.size());
+        lastChunk = lastChunk.concat(ByteString.copyFrom(buffer, bytesToWrite));
+      } else {
+        bytesToWrite = min(buffer.remaining(), MAX_BYTES_PER_MESSAGE);
+        lastChunk = ByteString.copyFrom(buffer, bytesToWrite);
+      }
+      bytesWritten += bytesToWrite;
+      if (lastChunk.size() == MAX_BYTES_PER_MESSAGE) {
+        try {
+          writeBuffer.put(lastChunk);
+          lastChunk = null;
+        } catch (InterruptedException e) {
+          throw new IOException(
+              String.format(
+                  "Failed to write %d bytes in '%s'", buffer.remaining(), getResourceString()),
+              e);
+        }
+      }
+    }
+
+    return bytesWritten;
+  }
+
+  /** Initialize this channel object for writing. */
+  @Override
+  public void initialize() throws IOException {
+    checkState(!initialized, "initialize() must be invoked only once.");
+    writeBuffer =
+        new LinkedBlockingQueue<ByteString>(
+            Math.toIntExact(channelOptions.getNumberOfBufferedRequests()));
     try {
-      uploadOperation = threadPool.submit(new UploadOperation(pipeSource));
+      uploadOperation = threadPool.submit(new UploadOperation(writeBuffer));
     } catch (Exception e) {
       throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
     }
+    initialized = true;
+  }
+
+  @Override
+  public void close() throws IOException {
+    checkState(initialized, "initialize() must be invoked before use.");
+    if (!isOpen()) {
+      return;
+    }
+    // The way we signal close (EOF) to downstream thread is by appending a ByteString of size less
+    // than 2MB. If the uploaded file size is multiple of 2MB, then we end up appending an empty
+    // ByteString buffer. The lastChunk carries the data written by upstream not amounting to 2MB
+    if (lastChunk == null) {
+      lastChunk = ByteString.EMPTY;
+    }
+    try {
+      writeBuffer.put(lastChunk);
+      lastChunk = null;
+      handleResponse(waitForCompletionAndThrowIfUploadFailed());
+    } catch (InterruptedException e) {
+      throw new IOException(String.format("Interrupted on close for '%s'", resourceId), e);
+    } finally {
+      writeBuffer = null;
+      closeInternal();
```

Comment:
this is dependency to base class is not clean, we need to find a better way to achieve the same behaviour